### PR TITLE
Add `preferredPosition` to polaris-popover

### PR DESCRIPTION
**Pull request by tomnez**
## Overview

This adds the `preferredPosition` to the [polaris-popover component](https://polaris.shopify.com/components/overlays/popover).

It basically translates the user-provided `preferredPosition` value into a `verticalPosition` attribute since we use [ember-basic-dropdown](http://ember-basic-dropdown.com/) as our popover component under the hood.

Test for this is skipped because ember-basic-dropdown uses ember wormhole for the popover content, which renders content outside of the ember testing container.

### CODE REVIEW

File: addon/components/polaris-popover.js [hunk 1]
```diff
@@ -90,11 +93,73 @@ export default Component.extend({
...
+
+    if (preferredPosition === 'below') {
+      return preferredPosition;
+    }
+
+    // If set to `mostSpace`, the value will be calculated
+    // and set when the user opens the popover.
+
+    return null;
+  }),
```
Comment (andrewpye) [inline addon/components/polaris-popover.js]: `.readOnly()`?

File: addon/components/polaris-popover.js [hunk 2]
```diff
@@ -90,11 +93,73 @@ export default Component.extend({
...
+  verticalPosition: computed('preferredPosition', function() {
+    let preferredPosition = this.get('preferredPosition');
+
+    if (preferredPosition === 'above') {
+      return preferredPosition;
+    }
+
+    if (preferredPosition === 'below') {
+      return preferredPosition;
+    }
```
Comment (andrewpye) [inline addon/components/polaris-popover.js]: Why not bundle these two into a single `if`-statement?

File: addon/components/polaris-popover.js [hunk 3]
```diff
@@ -90,11 +93,73 @@ export default Component.extend({
...
+  /**
+   * Checks the dropdown trigger's location on
+   * screen to determine which vertical direction
+   * has more space to open the dropdown.
+   */
+  mostVerticalSpace() {
+    let component;
+
+    // Hack to get the component's container
+    // element since this component uses `tagName: ''`
```
Comment (andrewpye) [inline addon/components/polaris-popover.js]: Add a link here to the RFC I linked you to so we know where this came from in the future 👍

File: addon/components/polaris-popover.js [hunk 4]
```diff
@@ -90,11 +93,73 @@ export default Component.extend({
...
+
+  actions: {
+    onOpen() {
+      // Check to see if `preferredPosition` is set to `mostSpace`
+      // onOpen, since user could have scrolled vertically since
+      // the time the component originally rendered.
+      let preferredPosition = this.get('preferredPosition');
+
+      if (preferredPosition === 'mostSpace') {
+        this.set('verticalPosition', this.mostVerticalSpace());
```
Comment (andrewpye) [inline addon/components/polaris-popover.js]: Not a huge fan of the naming of `mostVerticalSpace` as a method - it sounds more like a property since there's no verb in the name. How about `getMostVerticalSpace` or making it a `volatile` CP and keeping the `mostVerticalSpace` name?

File: docs/popover.md [hunk 5]
```diff
@@ -6,7 +6,7 @@
 component](https://polaris.shopify.com/components/overlays/popover). This component uses [`ember-basic-dropdown`](https://github.com/cibernox/ember-basic-dropdown) to implement popover functionality. Note that the usage is slightly different from the React implementation so please pay attention to the examples below.
 
 
-**NOTE:** _the `preferredPosition`, `active`, `activatorWrapper`, `preventAutofocus` and `onClose` properties are currently unimplemented._
+**NOTE:** _the `active`, `activatorWrapper`, `preventAutofocus` and `onClose` properties are currently unimplemented._
```
Comment (andrewpye) [inline docs/popover.md]: 👍

File: tests/integration/components/polaris-popover-test.js [hunk 6]
```diff
@@ -101,3 +103,29 @@ test('it renders the correct HTML with sectioned attribute', function(assert) {
   assert.equal(popoverSections.length, 1, 'renders one popover section');
   assert.equal(popoverSections[0].textContent.trim(), 'This is some sectioned popover content', 'popover section contains the correct content');
 });
+
+skip('it renders the correct HTML with `prefferedPosition` attribute', function(assert) {
```
[the remark this answers is not in the document]
Comment (vladucu) [reply tests/integration/components/polaris-popover-test.js]: yes, I'd lean towards using `document.querySelector` (or whatever works) and not skip the test if possible

File: addon/components/polaris-popover.js (hunk 3 above)
Comment (vladucu) [reply addon/components/polaris-popover.js]: good with me for now, ideally I think we should investigate refactoring this at some point and not using `ember-basic-dropdown`

File: addon/components/polaris-popover.js (hunk 1 above)
Comment (tomnez) [reply addon/components/polaris-popover.js]: This gets `set` by the `onOpen` action if the position is set to `"mostSpace"`, so it can't be readOnly

File: tests/integration/components/polaris-popover-test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tomnez) [reply tests/integration/components/polaris-popover-test.js]: 👍 I'll see if `document.querySelector` works

File: addon/components/polaris-popover.js (hunk 1 above)
Comment (andrewpye) [reply addon/components/polaris-popover.js]: Hmm... seems like an _unlikely_ scenario, but `set`ting this in `onOpen` will mean that if the host app passes in a dynamic `preferredPosition` and that value changes from 'mostSpace' after first open, the popover won't honour that change because `onOpen` will overwrite the CP behaviour here. You _could_ add an explicit setter to this CP, and that way I think that (really tiny) edge case would be avoided...

**Comment by tomnez** [comment]
@vladucu @andrewpye updated - also I left the `controllers/test/index.js` file in there because I have found myself repeatedly re-creating it to test new components. Any opposition to this?

File: addon/components/polaris-popover.js (hunk 1 above)
Comment (tomnez) [reply addon/components/polaris-popover.js]: `onOpen` doesn't overwrite anything. `verticalPosition` is still a CP, so if `preferredPosition` dynamically changes, it will re-set `verticalPosition` to the new value, and then when `onOpen` runs it will _not_ compute a new value for `verticalPosition` because it only computes a new value if `preferredPosition` is set to `"mostSpace"`.

Comment (andrewpye) [reply addon/components/polaris-popover.js]: After `set`ting `verticalPosition` on [line 161](https://github.com/smile-io/ember-polaris/blob/3915ec85eb771e0441db38b6de7f8c5d8762c9c3/addon/components/polaris-popover.js#L161), my understanding is that `verticalPosition` will _no longer be a CP_ - the `set` will replace the entire CP with the literal value that you're setting it to, unless the CP has its own `set` defined as well. Think of it like this, where `someObject` is the result of the `computed` call:

```js
  ... other props...

  myComputed: someObject,

  ... other stuff...
```

If you then `this.set('myComputed', 'something else')`, you'll _replace `someObject` with the new string literal_. The CP "contained" in `someObject` will be torn down, and will no longer be computed on dependency changes. From then on, `this.get('myComputed')` will _always_ return `'something else'`.

You can work around this by defining the CP as

```js
myComputed: computed('dep1', 'dep2', {
  get() {
    return 'whatever';
  },

  set(key, value) {
    return value;
  }
),
```

The `set` part ensures that the CP remains after a `this.set('myComputed', 'something else')`, and the value will still update after `dep1` or `dep2` change...

File: addon/templates/components/polaris-popover.hbs [hunk 7]
```diff
@@ -1,4 +1,6 @@
 {{#basic-dropdown
+  verticalPosition=verticalPosition
+  onOpen=(action 'onOpen')
```
Comment (andrewpye) [inline addon/templates/components/polaris-popover.hbs]: Double-quotes in template files pls 😉

File: tests/integration/components/polaris-date-picker-test.js [hunk 8]
```diff
@@ -339,7 +339,7 @@ test('it does not fire actions when disabled days are clicked', function(assert)
     year: YEAR,
     selected: null,
     disableDatesAfter: DISABLE_AFTER,
-    onChange: (selected) => {
+    onChange: () => {
```
Comment (andrewpye) [inline tests/integration/components/polaris-date-picker-test.js]: Not sure you need the colon here actually

File: tests/integration/components/polaris-popover-test.js [hunk 9]
```diff
@@ -101,3 +103,40 @@ test('it renders the correct HTML with sectioned attribute', function(assert) {
...
+  // close the popover before resetting position
+  click(activatorSelector);
+
+  this.set('preferredPosition', 'below');
+
+  click(activatorSelector);
+
+  const contentBelow = document.querySelector(popoverContentBelowSelector);
+  assert.ok(contentBelow, 'renders popover content below the trigger');
+});
```
Comment (andrewpye) [inline tests/integration/components/polaris-popover-test.js]: Love that you got this working 😍
